kstatus/wait/wait.go

```diff
@@ -0,0 +1,222 @@
+				})
+			}
+			continue
+		}
+		res, err := r.statusComputeFunc(u)
+		results = append(results, ResourceResult{
+			Result:   res,
+			Resource: resource,
+			Error:    err,
+		})
+	}
+
+	return results
+}
+
+// Event is returned through the channel returned after a call
+// to WaitForStatus. It contains an update to either an individual
+// resource or to the aggregate status for the set of resources.
+type Event struct {
+	// Type defines which type of event this is.
+	Type EventType
+
+	// AggregateStatus is the aggregated status for all the provided resources.
+	AggregateStatus status.Status
+
+	// EventResource is information about the event to which this event pertains.
+	// This is only populated for ResourceUpdate events.
+	EventResource *EventResource
+}
+
+type EventType string
+
+const (
+	// The status/message for a resource has changed. This also means the
+	// aggregate status might have changed.
+	ResourceUpdate EventType = "ResourceUpdate"
+
+	// All resources has reached the current status.
+	Completed EventType = "Completed"
+
+	// The wait was stopped before all resources could reach the
+	// Current status.
+	Aborted EventType = "Aborted"
+)
+
+// EventResource contains information about the resource for which
+// a specific Event pertains.
+type EventResource struct {
+	// Identifier contains information that identifies which resource
+	// this information is about.
+	Identifier ResourceIdentifier
+
+	// Status is the latest status for the given resource.
+	Status status.Status
+
+	// Message is more details about the status.
+	Message string
+
+	// Error is set if there was a problem identifying the status
+	// of the resource. For example, if polling the cluster for information
+	// about the resource failed.
+	Error error
+}
+
+// WaitForStatus polls all the provided resources until all of them has
+// reached the Current status. Updates as resources change their status and
+// when the wait is either completed or aborted is returned through a channel.
+func (r *Resolver) WaitForStatus(ctx context.Context, resources []ResourceIdentifier, pollInterval time.Duration) <-chan Event {
+	eventChan := make(chan Event)
+
+	go func() {
+		ticker := time.NewTicker(pollInterval)
+		defer func() {
+			ticker.Stop()
+			close(eventChan)
+		}()
+
+		if len(resources) == 0 {
+			eventChan <- Event{
+				Type:            Completed,
+				AggregateStatus: status.CurrentStatus,
+				EventResource:   nil,
+			}
+			return
+		}
+
+		waitState := newWaitState(resources, r.statusComputeFunc)
+
+		for {
+			select {
+			case <-ctx.Done():
+				eventChan <- Event{
+					Type:            Aborted,
+					AggregateStatus: waitState.AggregateStatus(),
+				}
+				return
+			case <-ticker.C:
+				for id := range waitState.ResourceWaitStates {
+					identifier := id
+					u, err := r.fetchResource(ctx, &identifier)
+					eventResource, isNew := waitState.ResourceObserved(&identifier, u, err)
+					aggStatus := waitState.AggregateStatus()
```

Comment:
I am guessing this is to check if the most recent updated made the aggregated status done.  We should document this with comments, upon first looking at it I would assume this would be done outside the loop.